src/integrationTest/java/org/opensearch/test/framework/ldap/LdapServer.java

```diff
@@ -0,0 +1,250 @@
+import java.io.StringReader;
+import java.net.BindException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.ReentrantLock;
+
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+
+import com.unboundid.ldap.listener.InMemoryDirectoryServer;
+import com.unboundid.ldap.listener.InMemoryDirectoryServerConfig;
+import com.unboundid.ldap.listener.InMemoryListenerConfig;
+import com.unboundid.ldap.sdk.DN;
+import com.unboundid.ldap.sdk.Entry;
+import com.unboundid.ldap.sdk.LDAPException;
+import com.unboundid.ldap.sdk.schema.Schema;
+import com.unboundid.ldif.LDIFReader;
+import com.unboundid.util.ssl.SSLUtil;
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import org.opensearch.test.framework.certificate.CertificateData;
+import org.opensearch.test.framework.cluster.SocketUtils;
+
+/**
+* Based on class com.amazon.dlic.auth.ldap.srv.LdapServer from older tests
+*/
+final class LdapServer {
+	private static final Logger log = LogManager.getLogger(LdapServer.class);
+
+	private static final int LOCK_TIMEOUT = 60;
+	private static final TimeUnit TIME_UNIT = TimeUnit.SECONDS;
+
+	private static final String LOCK_TIMEOUT_MSG = "Unable to obtain lock due to timeout after " + LOCK_TIMEOUT + " " + TIME_UNIT.toString();
+	private static final String SERVER_NOT_STARTED = "The LDAP server is not started.";
+	private static final String SERVER_ALREADY_STARTED = "The LDAP server is already started.";
+
+	private final CertificateData trustAnchor;
+
+	private final CertificateData ldapCertificate;
+
+	private InMemoryDirectoryServer server;
+	private final AtomicBoolean isStarted = new AtomicBoolean(Boolean.FALSE);
+	private final ReentrantLock serverStateLock = new ReentrantLock();
+
+	private int ldapPort = -1;
+	private int ldapsPort = -1;
+
+
+	public LdapServer(CertificateData trustAnchor, CertificateData ldapCertificate) {
+		this.trustAnchor = trustAnchor;
+		this.ldapCertificate = ldapCertificate;
+	}
+
+	public boolean isStarted() {
+		return this.isStarted.get();
+	}
+
+	public int getLdapPort() {
+		return ldapPort;
+	}
+
+	public int getLdapsPort() {
+		return ldapsPort;
+	}
+
+	public void start(LdifData ldifData) throws Exception {
+		Objects.requireNonNull(ldifData, "Ldif data is required");
+		boolean hasLock = false;
+		try {
+			hasLock = serverStateLock.tryLock(LdapServer.LOCK_TIMEOUT, LdapServer.TIME_UNIT);
```

Comment:
does this lock lock some CPU cycles? If not, what role does this lock play here?

Reply:
Actually, I reused this class from older tests. I suppose that the lock is employed to prevent parallel start method execution so that the server can be started only once. The lock should not cause any performance-related issues.